### DISCUSSION TITLE
fast 3d libfive tree slicing algorithm

### DISCUSSION THREAD

**Discussion by wangsong004** [libfive/libfive#576] [Ideas]
"I'm performing slicing operations on a three-dimensional libfive::Tree. The fastest method I can think of is to call Contours::render layer by layer and use OpenMP for parallelization, but I still find it slow. Because I think you need to build the tree independently for each layer. This image is my code. 

![image](https://github.com/libfive/libfive/assets/166193797/4aec104e-0227-4303-9922-870ed9cb1316)

 In many cases, the layers are highly similar with only slight differences. Is it possible to refer to the situation of layer n when building the tree for layer n+1? Or build trees for multiple layers, such as 16 or 32 layers, simultaneously? Would this be faster?"

**Comment by mkeeter** [comment]
It's possible to pass a `VolTree` into contour rendering, which reuses a volume tree for faster slicing.

See https://github.com/libfive/libfive/blob/302553e6aa6ca3cb13b2a149f57b6182ce2406dd/libfive/test/vol_tree.cpp#L52 as an example

**Reply by wangsong004** [reply]
![image](https://github.com/libfive/libfive/assets/166193797/6b74aeec-2efa-4d09-a3eb-37898dc6a104)
![image](https://github.com/libfive/libfive/assets/166193797/f9d41b46-006f-4911-86b2-ea87308646f5)
The first image took me 3.196s,the second 3.373s. There hasn't been a significant change in speed I guess
